README.md
## Snake in rust using [ncurses](https://docs.rs/ncurses) crate

The game is playable

run with

```bash
cargo run
```

#### Keybinds

hjkl or arrow keys for movement

esc or p to pause

#### Screenshot

Click the image for asciinema

[<img src="images/screenshot.png" width="650" />](https://asciinema.org/a/FVB1DZmA7lVK4BILFkzkreMQ9?autoplay=1)

#### Need unicode support in terminal to print the snake

The unicode characters used are:

- 0x0298 ʘ
- 0x2550 ═
- 0x2551 ║
- 0x2554 ╔
- 0x2557 ╗
- 0x255a ╚
- 0x255d ╝

#### Todo :construction:

- [ ] Add a way to change the snake speed
  - [ ] Make a ui to switch the snake speed
  - [x] Interal implementation of the snake speed
- [ ] Implement Highscore System
  - [ ] Make the ui
  - [ ] Internal Implementation
- [x] Use unicode charachters to draw the snake

#### Bugs :bug:

- ~~Snake going through the walls~~
- ~~Food spawning in the walls~~
- ~~Boxdraw characters not rendering properly~~
- Remove all the logging in the ui
- Pausing delayed if esc is pressed but not if p is pressed.
- The game flickers a lot after sometime (probably due to the constant redrawing of the whole board every tick)

#### Maybe in the future

- [ ] Autoplay the game using a simple pathfinding algorithm (to show on [r/unixporn](https://reddit.com/r/unixporn) of course :clown_face:)

#### Notes

A few notes about the game and how I should improve the game

<details>
<summary>A few notes about time complexity and redrawing with ncurses</summary>

> The complexity of the program is O(n) every tick (time which changes relative to the speed)

> However the place, I can improve is the redrawing of the game

> At the time of writing the in commit
> <a href="https://github.com/uttarayan21/snake/commit/de66f7d249a56f883dd632598a4178b1bd1320ba">f9be68e</a>
> the game still redraws the total board and the total snake every tick.

> I think this can be improved by only drawing the parts of the snake and the board when needed

</details>
<br>

<details>

<summary>Rust specific stuff I learned</summary>

> [char](https://doc.rust-lang.org/stable/std/primitive.char.html) and [std::char](https://doc.rust-lang.org/stable/std/char/index.html) are not the same.
> [char](https://doc.rust-lang.org/stable/std/primitive.char.html) is the primitive type [std::char](https://doc.rust-lang.org/stable/std/char/index.html) is the char module.

> As of Sat, 23 Jan 2021 21:49:13 +0000 the char::from_u32 (used in this program) is not valid in primitive [char](https://doc.rust-lang.org/stable/std/primitive.char.html) but is valid in [std::char](https://doc.rust-lang.org/stable/std/char/index.html)

> [ncursesw](httsp://docs.rs/ncursesw) is not needed to print unicode characters. I have no clue where I got that idea from.

</details>


docs/CHANGELOG.md
- v0.2.1

  - > Change all the i32 variables to u32.
  - > Changed the snake logic a bit so that the snake doesn't clear the whole board for every draw


src/main.rs
extern crate ncurses;
mod game;
mod highscore;
mod menu;
// use game::{Cell, Snake};
// use ncurses::*;
use ncurses::{
    curs_set, endwin, getmaxyx, initscr, keypad, noecho, raw, refresh, setlocale, stdscr,
    LcCategory, CURSOR_VISIBILITY,
};

fn main() {
    // let (lines, cols): (i32, i32) = (0, 0);
    setlocale(LcCategory::all, "");
    initscr();
    raw();
    keypad(stdscr(), true);
    curs_set(CURSOR_VISIBILITY::CURSOR_INVISIBLE);
    noecho();
    let (mut mlines, mut mcols): (i32, i32) = (0, 0);
    getmaxyx(stdscr(), &mut mlines, &mut mcols);
    if (mlines < 20) || (mcols < 35) {
        refresh();
        endwin();
        eprintln!("Sorry window size too small");
        std::process::exit(1);
    }
    loop {
        match menu::main_menu_control() {
            0 => game::start(),
            1 => highscore::show(),
            _ => break,
        }
    }
    refresh();
    endwin();
}


src/menu.rs
use ncurses::*;
fn menu_window(mlines: i32, mcols: i32, lines: i32, cols: i32) -> WINDOW {
    let menu_win: WINDOW;
    let (starty, startx): (i32, i32);
    starty = (mlines - lines) / 2;
    startx = (mcols - cols) / 2;
    menu_win = newwin(lines, cols, starty, startx);
    box_(menu_win, 0, 0); // add default border
    wrefresh(menu_win); // refreshes the new window to draw it
    keypad(menu_win, true); // to use arrow keys
    menu_win
}
fn destroy_window(win: WINDOW) {
    wmove(win, 0, 0);
    wclrtobot(win);
    // wborder(win, 32, 32, 32, 32, 32, 32, 32, 32); // 32 is the ascii code for whitespace. this replaces all the window borders with whitespace
    wrefresh(win); // refresh to remove the borders
    delwin(win); // delete the window
}
pub fn pause_menu_control() -> i8 {
    let mut ch: i32 = 0;
    let choice: i8;
    let (mut lines, mut cols): (i32, i32) = (0, 0);
    let menu_win: WINDOW;
    let menu_items = vec!["Resume", "Restart", "Exit"];
    let menu_desc = vec!["Resume the game", "Restart the game", "Exit Game"];
    let mut menu_highlight: usize = 0;

    getmaxyx(stdscr(), &mut lines, &mut cols);
    menu_win = menu_window(lines, cols, 10, 20);
    loop {
        for (menu_item_num, menu_item) in menu_items.iter().enumerate() {
            if menu_item_num == menu_highlight {
                wattron(menu_win, A_REVERSE());
            }
            mvwaddstr(menu_win, menu_item_num as i32 + 1, 1, &menu_item);
            wattroff(menu_win, A_REVERSE());
            wrefresh(menu_win);
        }
        mvwaddstr(menu_win, 7, 1, "                  ");
        mvwaddnstr(menu_win, 7, 1, &menu_desc[menu_highlight], 18);
        mvwaddstr(menu_win, 8, 1, &format!("{}", ch));
        ch = wgetch(menu_win);
        match ch {
            KEY_UP | 107 => {
                if menu_highlight > 0 {
                    menu_highlight -= 1
                }
            }
            KEY_DOWN | 106 => {
                if menu_highlight < 2 {
                    menu_highlight += 1
                }
            }

            10 => {
                choice = menu_highlight as i8;
                break;
            }
            27 => {
                choice = 2;
                break;
            }
            _ => (),
        }
    }
    destroy_window(menu_win);
    choice
}
pub fn main_menu_control() -> i8 {
    let mut ch: i32 = 0;
    let choice: i8;
    let (mut lines, mut cols): (i32, i32) = (0, 0);
    let menu_win: WINDOW;
    let menu_items = vec!["Start", "High Score", "Exit"];
    let menu_desc = vec!["Start the game", "Check Highscrore", "Exit Game"];
    let mut menu_highlight: usize = 0;

    getmaxyx(stdscr(), &mut lines, &mut cols);
    menu_win = menu_window(lines, cols, 10, 20);
    loop {
        for (menu_item_num, menu_item) in menu_items.iter().enumerate() {
            if menu_item_num == menu_highlight {
                wattron(menu_win, A_REVERSE());
            }
            mvwaddstr(menu_win, menu_item_num as i32 + 1, 1, &menu_item);
            wattroff(menu_win, A_REVERSE());
            wrefresh(menu_win);
        }
        // mvwaddstr(menu_win, 7, 1, "                  ");
        wmove(menu_win, 7, 1);
        wclrtoeol(menu_win);
        box_(menu_win, 0, 0);
        mvwaddnstr(menu_win, 7, 1, &menu_desc[menu_highlight], 18);
        mvwaddstr(menu_win, 8, 1, &format!("{}", ch));
        ch = wgetch(menu_win);
        match ch {
            KEY_UP | 107 => {
                if menu_highlight > 0 {
                    menu_highlight -= 1
                }
            }
            KEY_DOWN | 106 => {
                if menu_highlight < 2 {
                    menu_highlight += 1
                }
            }
            10 => {
                choice = menu_highlight as i8;
                break;
            }
            27 => {
                choice = 2;
                break;
            }
            _ => (),
        }
    }
    destroy_window(menu_win);
    choice
}


src/game/mod.rs
mod backend;
mod frontend;
use crate::menu;
// use ncurses::*;
use backend::{Board, Cell, Snake};
use ncurses::{
    getmaxyx, nodelay, stdscr, wgetch, wrefresh, KEY_DOWN, KEY_LEFT, KEY_RIGHT, KEY_UP, WINDOW,
};
pub fn start() {
    let (mut mlines, mut mcols): (i32, i32) = (0, 0);
    let game_win: WINDOW;
    let mut ch: i32;
    let (vmargin, hmargin): (u32, u32) = (5, 10);
    getmaxyx(stdscr(), &mut mlines, &mut mcols);

    game_win = frontend::game_window(mlines as u32, mcols as u32, vmargin, hmargin);
    let mut snake = Snake::new(Cell::new(
        mlines as u32 / 2 - vmargin,
        mcols as u32 / 2 - hmargin,
        backend::CellType::Snake,
    )); //Initialise snake in the middle of the screen
    let mut board = Board::new(mlines as u32 - vmargin * 2, mcols as u32 - hmargin * 2);
    nodelay(game_win, true);
    loop {
        frontend::draw_snake(&snake, game_win); // always draw snake before board because the snake will clear the game win
        frontend::draw_board(&board, game_win);
        frontend::_log(&snake, &board);
        if board.check_collision(&snake) {
            // Add stuff here to show the score and
            // how You lose screen
            break;
        }
        if board.check_food(&snake) {
            // snake.grow();
            board.eat_food(&mut snake);
        }
        ch = wgetch(game_win);
        match ch {
            KEY_UP | 107 => snake.smove(backend::Direction::Up),
            KEY_DOWN | 106 => snake.smove(backend::Direction::Down),
            KEY_LEFT | 104 => snake.smove(backend::Direction::Left),
            KEY_RIGHT | 108 => snake.smove(backend::Direction::Right),
            112 | 27 => {
                nodelay(game_win, false);
                match menu::pause_menu_control() {
                    //112 is keycode for 'p'
                    0 => (), //resume
                    1 => {
                        snake = Snake::new(Cell::new(
                            mlines as u32 / 2,
                            mcols as u32 / 2,
                            backend::CellType::Snake,
                        )); //Initialise snake in the middle of the screen
                        board = Board::new(mlines as u32 - vmargin * 2, mcols as u32 - hmargin * 2);
                    } //restart
                    2 => break, //exit
                    _ => (), //other charachters just in case
                }
                wrefresh(game_win);
                nodelay(game_win, true);
            }
            // 27 => break,
            _ => snake.tick(),
        }
    }

    frontend::destroy_window(game_win);
}


src/game/backend.rs
extern crate rand;
use core::iter::Iterator;
use rand::Rng;
use std::collections::LinkedList;
use std::ops::Sub;
use std::thread::sleep;
use std::time::Duration;
#[derive(Debug)]
pub enum Direction {
    Up,
    Down,
    Left,
    Right,
}
impl Copy for Direction {}
impl Clone for Direction {
    fn clone(&self) -> Self {
        *self
    }
}
impl PartialEq for Direction {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (&Self::Up, &Self::Up) => true,
            (&Self::Down, &Self::Down) => true,
            (&Self::Left, &Self::Left) => true,
            (&Self::Right, &Self::Right) => true,
            _ => false,
        }
    }
}

impl Direction {
    fn opposite(&self) -> Self {
        match *self {
            Direction::Up => Direction::Down,
            Direction::Down => Direction::Up,
            Direction::Left => Direction::Right,
            Direction::Right => Direction::Left,
        }
    }
}
#[derive(Debug)]
// impl std::fmt::Debug for Direction {
//     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
//         // write!(f,)
//         write!(
//             f,
//             "{}",
//             match *self {
//                 Self::Up => (),
//                 _ => (),
//             }
//         );
//     }
// }

pub enum CellType {
    Food,
    Snake,
    Empty,
}
pub struct Cell {
    line: u32,
    col: u32,
    ctype: CellType,
}
impl PartialEq for Cell {
    fn eq(&self, other: &Cell) -> bool {
        if (self.line == other.line) && (self.col == other.col) {
            return true;
        } else {
            return false;
        }
    }
}
impl Sub for Cell {
    type Output = (i32, i32);
    fn sub(self, rhs: Cell) -> (i32, i32) {
        return (
            self.line as i32 - rhs.line as i32,
            self.col as i32 - rhs.col as i32,
        );
    }
}
impl Cell {
    pub fn new(l: u32, c: u32, t: CellType) -> Cell {
        Cell {
            line: l,
            col: c,
            ctype: t,
        }
    }
    pub fn random(lines: u32, cols: u32) -> Cell {
        let mut rng = rand::thread_rng();
        Cell::new(
            rng.gen_range(1..lines - 1) as u32,
            rng.gen_range(1..cols - 1) as u32,
            CellType::Food,
        )
    }
    pub fn posyx(&self) -> (u32, u32) {
        return (self.line, self.col);
    }
    pub fn posy(&self) -> u32 {
        return self.line;
    }
    pub fn posx(&self) -> u32 {
        return self.col;
    }
    pub fn chtype(&mut self, ctype: CellType) {
        self.ctype = ctype;
    }
    pub fn is_adjacent(&self, other: &Cell) -> Option<Direction> {
        match *self - *other {
            (0, 1) => Some(Direction::Right),
            (1, 0) => Some(Direction::Down),
            (-1, 0) => Some(Direction::Up),
            (0, -1) => Some(Direction::Left),
            _ => None,
        }
    }
}

impl Copy for CellType {}
impl Clone for CellType {
    fn clone(&self) -> CellType {
        *self
    }
}
impl Copy for Cell {}
impl Clone for Cell {
    fn clone(&self) -> Cell {
        *self
    }
}

enum FailState {
    Wall,
    Snake,
}

enum GameState {
    Failed(FailState),
    Ready,
    // Playing,
}

pub struct Board {
    maxlines: u32,
    maxcols: u32,
    gamestate: GameState,
    food: Cell,
}

// impl Board<'_> {
impl Board {
    pub fn new(maxlines: u32, maxcols: u32) -> Board {
        Board {
            maxlines,
            maxcols,
            gamestate: GameState::Ready,
            food: Cell::random(maxlines, maxcols),
        }
    }
    pub fn check_collision(&mut self, snake: &Snake) -> bool {
        let (snake_line, snake_col): (u32, u32) = snake.posyx();
        if (snake_line >= self.maxlines - 1)
            || (snake_col >= self.maxcols - 1)
            || (snake_line <= 0)
            || (snake_col <= 0)
        {
            self.gamestate = GameState::Failed(FailState::Wall);
            return true;
        }
        let mut snake_iter = snake.iter();
        snake_iter.next();
        for snake_cell in snake_iter {
            // O(n) ; don't know how to reduce this complexity
            if snake.posyx() == snake_cell.posyx() {
                self.gamestate = GameState::Failed(FailState::Snake);
                return true;
            }
        }
        return false;
    }
    pub fn check_food(&self, snake: &Snake) -> bool {
        if self.food_posyx() == snake.posyx() {
            return true;
        } else {
            return false;
        }
    }
    pub fn food_posyx(&self) -> (u32, u32) {
        return self.food.posyx();
    }
    // pub fn food_posy(&self) -> u32 {
    //     return self.food.line;
    // }
    pub fn eat_food(&mut self, snake: &mut Snake) {
        snake.grow();
        // self.food.chtype(CellType::Empty);
        self.spawn_food(snake);
    }
    pub fn spawn_food(&mut self, snake: &Snake) {
        let mut food: Cell = Cell::random(self.maxlines, self.maxcols);
        let mut spawned_food = false;
        while spawned_food != true {
            // check for colliosions with the snake body until a free spot is found and spawn the food there
            spawned_food = true;
            let mut snake_iter = snake.iter();
            for snake_cell in snake_iter.next() {
                if *snake_cell == food {
                    // O(n) ; I think this is nessacary/ I don't know how to reduce the order
                    // if food collides with the snake body then set food to a new random position and set spawned food to false
                    // so that the snake_iter is started again from the front of the snake
                    food = Cell::random(self.maxlines, self.maxcols);
                    spawned_food = false;
                    break;
                }
            }
        }
        self.food = food;
    }
}

pub struct Snake {
    head: Cell,
    body: LinkedList<Cell>,
    direction: Direction,
    grow: bool,
    speed: u32,
    last_tail: Option<Cell>,
}
impl Snake {
    pub fn new(head: Cell) -> Snake {
        let mut temp_body: LinkedList<Cell> = LinkedList::new();
        temp_body.push_front(head);
        Snake {
            head,
            body: temp_body,
            // length: 1,
            direction: Direction::Right,
            grow: false,
            speed: 15,
            last_tail: Some(head),
        }
    }
    pub fn posyx(&self) -> (u32, u32) {
        return (self.head.line, self.head.col);
    }
    pub fn smove(&mut self, _direction: Direction) {
        // smove because move is already a keyword
        let direction: Direction;
        if self.direction == _direction.opposite() {
            direction = self.direction
        } else {
            direction = _direction
        }
        if self.grow == false {
            self.last_tail = Some(self.body.pop_back().unwrap());

            // self.grow = false;
        }
        let (dl, dc): (i32, i32) = match direction {
            Direction::Up => (-1, 0),
            Direction::Down => (1, 0),
            Direction::Left => (0, -1),
            Direction::Right => (0, 1),
        };
        self.direction = direction;
        self.head = Cell::new(
            ((self.head.line as i32) + dl) as u32,
            ((self.head.col as i32) + dc) as u32,
            CellType::Snake,
        );
        self.body.push_front(self.head);
        self.grow = false;
    }
    pub fn tick(&mut self) {
        // sleep(time);
        // let time: std::time::Duration =
        //     std::time::Duration::from_millis((1000 / self.speed) as u64);
        // sleep(time);
        sleep(Duration::from_millis((1000 / self.speed) as u64));
        self.smove(self.direction);
    }
    pub fn grow(&mut self) {
        self.grow = true;
    }
    pub fn iter(&self) -> impl Iterator<Item = &Cell> {
        return self.body.iter();
    }
    pub fn remove(&self) -> Option<Cell> {
        return self.last_tail;
    }
    pub fn set_speed(&mut self, speed: u32) {
        self.speed = speed;
    }
}


src/game/frontend.rs
// use ncurses::*;
use crate::game::backend::{Board, Cell, Direction, Snake};
use ncurses::{
    box_, delwin, keypad, mvwaddstr, newwin, stdscr, waddstr, wborder, wclrtobot, wmove, wrefresh,
    WINDOW,
};
pub fn game_window(mlines: u32, mcols: u32, vmargin: u32, hmargin: u32) -> WINDOW {
    let game_win: WINDOW;
    let (lines, cols): (u32, u32);
    let (starty, startx): (u32, u32);
    lines = mlines - vmargin * 2;
    cols = mcols - hmargin * 2;
    starty = vmargin;
    startx = hmargin;
    game_win = newwin(lines as i32, cols as i32, starty as i32, startx as i32);
    box_(game_win, 0, 0);
    keypad(game_win, true);
    wrefresh(game_win);
    game_win
}

pub fn destroy_window(win: WINDOW) {
    wmove(win, 0, 0);
    wclrtobot(win);
    wborder(win, 32, 32, 32, 32, 32, 32, 32, 32); // 32 is the ascii code for whitespace. this replaces all the window borders with whitespace
    wrefresh(win); // refresh to remove the borders
    delwin(win); // delete the window
}

pub fn draw_snake(snake: &Snake, game_win: WINDOW) {
    // let mut snake_iter = snake.iter();
    let (mut prev, mut current, _next): (&Cell, &Cell, &Cell);
    let mut snake_iter = snake.iter();
    wmove(game_win, 0, 0);
    // wclrtobot(game_win);
    box_(game_win, 0, 0);

    // I want to draw the snake as ascii box charachters
    // So I'll need to know the last and next cell of the snake to draw the current snake_cell
    // For some reason the snake goes invisible after the first food
    prev = snake_iter.next().unwrap(); // currently this should be head. On initial run this should be the only snake_cell
    mvwaddstr(
        game_win,
        prev.posy() as i32,
        prev.posx() as i32,
        &format!("{}", std::char::from_u32(0x0298).unwrap_or('O')),
    );
    match snake.remove() {
        Some(tail) => {
            mvwaddstr(game_win, tail.posy() as i32, tail.posx() as i32, " ");
        }
        None => (),
    }
    let _current = snake_iter.next();
    current = match _current {
        Some(cell) => cell,
        None => return,
    };
    for next in snake_iter {
        // O(n) the whole snake is redrawn every single tick
        let (snake_l, snake_c): (u32, u32) = current.posyx();
        // mvwaddstr(game_win, snake_l, snake_c, "o");
        let snake_char: u32 = match (
            prev.is_adjacent(current).unwrap(),
            next.is_adjacent(current).unwrap(),
        ) {
            (Direction::Up, Direction::Down) | (Direction::Down, Direction::Up) => 0x2551, //"║"
            (Direction::Up, Direction::Left) | (Direction::Left, Direction::Up) => 0x255d, //"╝"
            (Direction::Up, Direction::Right) | (Direction::Right, Direction::Up) => 0x255a, //"╚"
            (Direction::Down, Direction::Left) | (Direction::Left, Direction::Down) => 0x2557, // "╗"
            (Direction::Down, Direction::Right) | (Direction::Right, Direction::Down) => 0x2554, //"╔"
            (Direction::Left, Direction::Right) | (Direction::Right, Direction::Left) => 0x2550, //"═"
            _ => 0x20,
        };
        mvwaddstr(
            game_win,
            snake_l as i32,
            snake_c as i32,
            &format!("{}", std::char::from_u32(snake_char).unwrap_or('o')),
        );
        prev = current;
        current = next;
    }

    mvwaddstr(
        game_win,
        current.posy() as i32,
        current.posx() as i32,
        &format!(
            "{}",
            std::char::from_u32(match current.is_adjacent(prev).unwrap() {
                Direction::Up | Direction::Down => 0x2551,
                Direction::Left | Direction::Right => 0x2550,
            })
            .unwrap_or('o')
        ),
    );
    wrefresh(game_win);
}

pub fn draw_board(board: &Board, game_win: WINDOW) {
    let (food_l, food_c): (u32, u32) = board.food_posyx();
    mvwaddstr(
        game_win,
        food_l as i32,
        food_c as i32,
        &format!("{}", std::char::from_u32(0x0298).unwrap_or('F')),
    );
}

pub fn _log(_snake: &Snake, _board: &Board) {
    let (shl, shc): (u32, u32) = _snake.posyx();
    let (stl, stc): (u32, u32) = _snake.remove().unwrap().posyx();
    mvwaddstr(stdscr(), 0, 0, &format!("snake:head: {} {} ", shl, shc));
    mvwaddstr(stdscr(), 1, 0, &format!("snake:tail: {} {} ", stl, stc));
    wmove(stdscr(), 2, 0);
    for snake_cell in _snake.iter() {
        waddstr(
            stdscr(),
            &format!(
                "<cell y:{} x:{}>",
                snake_cell.posyx().0,
                snake_cell.posyx().1
            ),
        );
    }
    wrefresh(stdscr());
}
